src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,4338 @@
+			(long int *)(&libc_close_nocancel));
+
+	register_a_hook("libc", "__read", (void *)new_read_libc, (long int *)(&libc_read));
+	register_a_hook("libpthread", "__read", (void *)new_read_pthread,
+			(long int *)(&pthread_read));
+	register_a_hook("libc", "__write", (void *)new_write_libc, (long int *)(&libc_write));
+	register_a_hook("libpthread", "__write", (void *)new_write_pthread,
+			(long int *)(&pthread_write));
+
+	register_a_hook("libc", "lseek64", (void *)new_lseek_libc, (long int *)(&libc_lseek));
+	register_a_hook("libpthread", "lseek64", (void *)new_lseek_pthread,
+			(long int *)(&pthread_lseek));
+
+	register_a_hook("libc", "unlink", (void *)new_unlink, (long int *)(&libc_unlink));
+
+	register_a_hook("libc", "__fxstat", (void *)new_fxstat, (long int *)(&next_fxstat));
+	register_a_hook("libc", "__xstat", (void *)new_xstat, (long int *)(&next_xstat));
+	/* Many variants for lxstat: _lxstat, __lxstat, ___lxstat, __lxstat64 */
+	register_a_hook("libc", "__lxstat", (void *)new_lxstat, (long int *)(&real_lxstat));
+	register_a_hook("libc", "__fxstatat", (void *)new_fxstatat, (long int *)(&next_fxstatat));
+	register_a_hook("libc", "readdir", (void *)new_readdir, (long int *)(&next_readdir));
+
+	register_a_hook("libc", "fcntl", (void *)new_fcntl, (long int *)(&libc_fcntl));
+
+	register_a_hook("libc", "mmap", (void *)new_mmap, (long int *)(&next_mmap));
+	register_a_hook("libc", "munmap", (void *)new_munmap, (long int *)(&next_munmap));
+
+	/**	register_a_hook("libc", "execve", (void *)new_execve, (long int *)(&real_execve));
+	 *	register_a_hook("libc", "execvp", (void *)new_execvp, (long int *)(&real_execvp));
+	 *	register_a_hook("libc", "execv", (void *)new_execv, (long int *)(&real_execv));
+	 *	register_a_hook("libc", "fork", (void *)new_fork, (long int *)(&real_fork));
+	 */
+
+	init_fd_dup2_list();
+	discover_daos_mount();
+
+	install_hook();
+
+	hook_enabled = 1;
+}
+
+static __attribute__((destructor)) void
+finalize_myhook(void)
+{
+	close_all_duped_fd();
+
+	uninstall_hook();
+	finalize_dfs();
+
+	pthread_mutex_destroy(&lock_init);
+	pthread_mutex_destroy(&lock_dirfd);
+	pthread_mutex_destroy(&lock_fd);
+	pthread_mutex_destroy(&lock_mmap);
+
+	free(file_list);
+	free(dir_list);
+}
+
+static void
+init_dfs(int idx)
+{
+	int rc;
+
+	/* TODO: Need to check the permission of mount point first!!! */
+	rc =
+	    daos_pool_connect(dfs_list[idx].pool, NULL, DAOS_PC_RW, &dfs_list[idx].poh, NULL, NULL);
+	if (rc != 0) {
+		printf("Failed to connect pool: %s\nQuit\n", strerror(rc));
+		exit(1);
+	}
+	rc = daos_cont_open(dfs_list[idx].poh, dfs_list[idx].cont, DAOS_COO_RW, &dfs_list[idx].coh,
+			    NULL, NULL);
+	if (rc != 0) {
+		printf("Failed to open container: %s\nQuit\n", strerror(rc));
+		exit(1);
+	}
+	rc = dfs_mount(dfs_list[idx].poh, dfs_list[idx].coh, O_RDWR, &dfs_list[idx].dfs);
+	if (rc != 0) {
+		printf("Failed to mount dfs: %s\nQuit\n", strerror(rc));
+		exit(1);
+	}
```

Comment:
can you just use dfs_init / dfs_connect in this case? this would also cache the pool and container handles for you and free them on dfs_fini.

Reply:
that makes sense. thanks